README.md
### Coisas a fazer e a estudar

#### Aba de configurações
Tem que se adicionar as configurações restantes.

#### Avisos de erro, não crashes
Fazer com que o programa não se autodestrua toda vez que algo dá levemente errado.


changelog.md
* Consertado formatação de CNPJ

neat.py
from numpy import trunc, float64, int64
from unidecode import unidecode
from locale import currency
from pandas import isna


def name(x) -> str:
    """
    Retorna o nome em CAIXA ALTA e sem acentos.
    """

    return unidecode(str(x).upper())


def capi(x) -> str:
    """
    Retorna o capital formatado para R$ XXX.XXX,XX
    """

    try:
        if isna(x) or x == 'ERRO' or x == '':
            x = 0.0

        # Isso só faz sentido assumindo que a localidade do arquivo Excel cujos dados foram extraídos está em pt-BR
        # Isso causará erros se isso não for verdadeiro.
        # todo Fazer isso de uma maneira mais independente
        elif isinstance(x, str):
            x = x.replace('.', '').replace('R', '').replace('$', '').strip()
            x = float(x.replace(',', '.'))

        elif isinstance(x, (float64, int64)):
            x = float(x)

    except ValueError:
        x = 0.0

    return currency(x, grouping=True)


def cpf(x) -> str:
    if isna(x):
        x = ''

    elif isinstance(x, (float64, int64, int, float)):
        x = str(int(trunc(x)))

    x = x.replace('.', '').replace('-', '').replace('/', '').replace('\\', '').strip()

    # Se tiver faltando números no CPF, adiciona zeros no começo.
    if len(x) < 11:
        x = '0' * (11 - len(x)) + x

    elif len(x) > 11:
        raise Exception(f'CPF "{x}" tem caracteres demais')

    # Formata e retorna o CPF.
    return x[:3] + '.' + x[3:6] + '.' + x[6:9] + '-' + x[9:]


def cnpj(x) -> str:
    if isna(x):
        x = ''

    elif isinstance(x, (float64, int64, int, float)):
        x = str(int(trunc(x)))

    x = x.replace('.', '').replace('-', '').replace('/', '').replace('\\', '').strip()

    # Se tiver faltando números no CNPJ, adiciona zeros no começo.
    if len(x) < 14:
        x = '0' * (14 - len(x)) + x

    elif len(x) > 14:
        raise Exception(f'CNPJ "{x}" tem caracteres demais')

    # Formata e retorna o CNPJ.
    return x[:2] + '.' + x[2:5] + '.' + x[5:8] + '/' + x[8:12] + '-' + x[12:]


def matr(x) -> str:
    if isna(x):
        x = ''

    elif isinstance(x, (float64, int64, int, float)):
        x = str(int(trunc(x)))

    if len(x) < 6:
        x = '0' * (6 - len(x)) + x  # Matrícula deve ter 6 caracteres.

    return x

def grupo(x) -> str:
    if isna(x):
        return 'GERENTES'

    elif isinstance(x, str):
        x = int(x.upper().replace('GRUPO', '').strip())
        x = str(x)  # Para se livrar de possíveis zeros na esquerda

    elif isinstance(x, (float64, int64, float, int)):
        x = str(int(trunc(x)))

    return x


parser.py
from unidecode import unidecode


def name(x):
    return unidecode(x.upper())


def cpf(x):
    # Transforma variável em string e retira pontos, hifens e traços se tiver.
    x = str(x).replace('.', '').replace('-', '').replace('/', '')

    # Se tiver faltando números no CPF, adiciona zeros no começo.
    if len(x) < 11:
        x = '0' * (11 - len(x)) + x

    elif len(x) > 11:
        raise Exception(f'CPF "{x}" tem caracteres demais')

    # Formata e retorna o CPF.
    return x[:3] + '.' + x[3:6] + '.' + x[6:9] + '-' + x[9:]


def cnpj(x):
    # Transforma variável em string e retira pontos, hifens e traços se tiver.
    x = str(x).replace('.', '').replace('-', '').replace('/', '')

    # Se tiver faltando números no CNPJ, adiciona zeros no começo.
    if len(x) < 14:
        x = '0' * (14 - len(x)) + x

    elif len(x) <= 15:
        x = x.replace('0', '', 1)

    else:
        raise Exception(f'CNPJ "{x}" tem caracteres demais')

    # Formata e retorna o CNPJ.
    return x[:2] + '.' + x[2:5] + '.' + x[5:8] + '/' + x[8:12] + '-' + x[12:]


def matr(x):
    x = str(x)

    if len(x) < 6:
        x = '0' * (6 - len(x)) + x

    elif len(x) > 6:
        raise Exception(f'Matrícula "{x}" tem caracteres demais')

    return x


def cnv(x):

    if x == 'GERENTES':
        return x

    x = str(int(x))

    if len(x) < 4:
        x = '0' * (4 - len(x)) + x

    elif len(x) > 4:
        raise Exception(f'CNV "{x}" tem caracteres demais')

    return x


reader.py
from pandas import read_excel, read_csv, DataFrame
from yaml import safe_load, safe_dump
from multiprocessing import Process
from unidecode import unidecode
from time import time
import var


# Função load usada exclusivamente para multiprocessamento. Em vez de retornar o valor, ele salva-o numa variável multiprocessamento no módulo var.
class SubLoader(Process):
    def __init__(self, path, skiprows, nrows, i, mlist, lock):
        super().__init__()
        self.path = path
        self.skiprows = skiprows
        self.nrows = nrows
        self.i = i
        self.list = mlist
        self.lock = lock
        self.loadtime = time()

    def run(self) -> None:
        if self.path[-4:] == "xlsx" or self.path[-4:] == ".xls" or self.path[-4:] == "xlsm":
            with self.lock:
                self.list[self.i] = read_excel(self.path, skiprows=self.skiprows, nrows=self.nrows, header=None)

        elif self.path[-4:] == ".csv":
            with self.lock:
                self.list[self.i] = read_csv(self.path, skiprows=self.skiprows, nrows=self.nrows, sep=';', header=None)

        print(f'Tempo levado para o {self.i + 1}º processo carregar {len(self.list[self.i].index)} linhas na memória: {(time() - self.loadtime):.2f} segundos.')


def load(path: str, **kwargs) -> DataFrame:
    """
    Função que distingue entre arquivos Excel e CSV, os lê e retorna um dataframe com o seu conteúdo. Frequentemente usado para ler apenas parte 
    do arquivo.
    """

    if path[-4:] == "xlsx" or path[-4:] == ".xls" or path[-4:] == "xlsm":
        return read_excel(path, **kwargs)

    elif path[-4:] == ".csv":
        return read_csv(path, sep=';', **kwargs)

def get_grupo(path: str) -> DataFrame:
    """
    Lê todos os códigos de cobertura num arquivo externo e retorna a relação cifras-coberturas num dataframe 
    (usado para a emissão de múltiplos certificados).
    """

    df = load(path)
    df = df.set_index('Grupo')
    df = df.squeeze()

    return df

def get_grupo_values(path: str) -> list:
    """
    Variável é um string que representa o caminho do diretório da tabela onde estão os grupos e retorna uma lista com todos os grupos válidos.
    """
    df = load(path)
    grupos = []
    from pandas import isna
    for i in range(len(df.index)):
        if not isna(df.iloc[i][1]):
            grupos.append(df.iloc[i][0])

    return grupos

def get_coberturas(path: str) -> list:
    """
    Lê um arquivo externo com todas as possibilidades de coberturas (usado para a emissão de certificados individuais).
    """

    df = load(path, header=None)
    cobe = []

    try:
        for e, value in enumerate(df.values):
            if e == 0: continue
            cobe.append(value[1])
    except IndexError:
        cobe = df.values

    return cobe


def get_headers(path: str) -> dict[str, str]:
    """
    Função que olha todos os cabeçários da planilha e compara os seus nomes com palavras-chave. Usado para descobrir onde estão os dados que 
    serão usados numa planilha desconhecida.
    """

    df = load(path, nrows=1)

    headers = {'name': '',
               'cpf' : '',
               'cnpj': '',
               'matr': '',
               'clie': '',
               'apol': '',
               'capi': '',
               'cobe': '',
               'grup': ''}

    # Depois que encontrar tal palavra-chave, ele não a substituirá se achar outra instância dela.
    nome_found = False
    cpf_found  = False
    cnpj_found = False
    matr_found = False
    clie_found = False
    apol_found = False
    capi_found = False
    cobe_found = False
    grup_found = False

    for column in df.columns:

        if any(key in unidecode(column.lower().strip()) for key in var.name_keywords) and not nome_found:  # Se qualquer chave na lista de chaves é contida nessa coluna, então é a coluna certa.
            headers['name'] = df.columns.get_loc(column)
            nome_found = True

        elif any(key in unidecode(column.lower().strip()) for key in var.cpf_keywords) and not cpf_found:  # Em todas essas iterações, lê se o nome da coluna em minísculo e se acentos.
            headers['cpf'] = df.columns.get_loc(column)
            cpf_found = True

        elif any(key in unidecode(column.lower().strip()) for key in var.cnpj_keywords) and not cnpj_found:
            headers['cnpj'] = df.columns.get_loc(column)
            cnpj_found = True

        elif any(key in unidecode(column.lower().strip()) for key in var.cliente_keywords) and not clie_found:
            headers['clie'] = df.columns.get_loc(column)
            clie_found = True

        elif any(key in unidecode(column.lower().strip()) for key in var.matricula_keywords) and not matr_found:
            headers['matr'] = df.columns.get_loc(column)
            matr_found = True

        elif any(key in unidecode(column.lower().strip()) for key in var.cobertura_keywords) and not cobe_found:
            headers['cobe'] = df.columns.get_loc(column)
            cobe_found = True

        elif any(key in unidecode(column.lower().strip()) for key in var.capital_keywords) and not capi_found:
            headers['capi'] = df.columns.get_loc(column)
            capi_found = True
        
        elif any(key in unidecode(column.lower().strip()) for key in var.grupo_keywords) and not grup_found:
            headers['grup'] = df.columns.get_loc(column)
            grup_found = True

        elif any(key in unidecode(column.lower().strip()) for key in var.apolice_keywords) and not apol_found:
            headers['apol'] = df.columns.get_loc(column)
            apol_found = True

    return headers  # Retorna um dicionário contendo os índices das colunas que contém os dados que serão usados.


def emergency_save() -> None:
    """
    Essa função salva todos os valores atuais.
    """

    with open('dados/init.yaml', 'w') as file:
        safe_dump({
            'max_threads'        :        var.max_threads,
            'max_processes'      :      var.max_processes,
            'target_threads'     :     var.target_threads,
            'data_dir'           :           var.data_dir,
            'output_dir'         :         var.output_dir,
            'start_period'       :       var.start_period,
            'end_period'         :         var.end_period,
            'cobertura_dir'      :      var.cobertura_dir,
            'template_dir'       :       var.template_dir,
            'text_font'          :          var.text_font,
            'text_size'          :          var.text_size,
            'dist_left'          :          var.dist_left,
            'text_height'        :        var.text_height,
            'line_space'         :         var.line_space,
            'max_chars'          :          var.max_chars,
            'base_text'          :          var.base_text,
            'name_keywords'      :      var.name_keywords,
            'cpf_keywords'       :       var.cpf_keywords,
            'cnpj_keywords'      :      var.cnpj_keywords,
            'matricula_keywords' : var.matricula_keywords,
            'cliente_keywords'   :   var.cliente_keywords,
            'apolice_keywords'   :   var.apolice_keywords,
            'capital_keywords'   :   var.capital_keywords,
            'cobertura_keywords' : var.cobertura_keywords,
            'grupo_keywords'     :     var.grupo_keywords,
        }, file)


def push(name: str, value) -> None:
    """
    Essa função salva um valor específico.
    """

    with open('dados/init.yaml', 'r') as file:
        dic = safe_load(file.read())

    dic[name] = value

    with open('dados/init.yaml', 'w') as file:
        safe_dump(dic, file)


if var.load_problem : emergency_save()  # Depois que os dois módulos estiverem carregados, se houve erro de carregamento, salve os novos valores.


emitter.py
from reportlab.pdfbase.pdfmetrics import registerFont
from reportlab.pdfbase.ttfonts import TTFont
from threading import active_count, Thread
from reportlab.pdfgen.canvas import Canvas
from pdfrw.toreportlab import makerl
from openpyxl import load_workbook
from time import sleep, time
from textwrap import wrap
from pandas import concat
from math import floor
from reader import *
import neat
import var


def emit_singular(name, cpf, cnpj, matr, clie, apol, capi, cobe='', grup='') -> None:
    """
    Método usado para emitir um certificado. Leva nome, CPF, CNPJ, matrícula, cliente, apólice e coberturas ou código. Retorna nada.
    """

    with var.lock:
        if var.progress == 0:
            # Configura a fonte a ser usada no PDF
            registerFont(TTFont(var.text_font, f'dados\\fontes\\{var.text_font}'))  # Registra essa fonte na memória

        var.progress += 1
        text = var.base_text  # Copia o texto base para uma variável nesse processo.

    # Substitui todos os campos dinâmicos pelos seus respectivos valores nessa instância.
    text = text.replace('<NOME>', neat.name(name))
    text = text.replace('<CPF>', neat.cpf(cpf))
    text = text.replace('<CNPJ>', neat.cnpj(cnpj))
    text = text.replace('<MATRICULA>', neat.matr(matr))
    text = text.replace('<CLIENTE>', neat.name(clie))
    text = text.replace('<APOLICE>', str(apol))
    text = text.replace('<COMECO>', f'{var.start_period.day:02}/{var.start_period.month:02}/{var.start_period.year}')
    text = text.replace('<FINAL>', f'{var.end_period.day:02}/{var.end_period.month:02}/{var.end_period.year}')
    text = text.replace('<CAPITAL>', neat.capi(capi))

    if cobe != '':
        text = text.replace('<COBERTURA>', cobe)

    else:
        text = text.replace('<COBERTURA>', str(grup))

    canvas = Canvas(f'{var.output_dir}{neat.cpf(cpf)} - {neat.name(name)}.pdf')  # Objeto do texto a ser salvo.
    canvas.setFont(var.text_font, var.text_size)  # Seleciona esse fonte para ser usada

    xobj_name = makerl(canvas, var.template_obj)  # Superposição objeto modelo com o objeto escritor por cima.
    canvas.doForm(xobj_name)

    userbreaks = text.split('\n')
    lines = []
    for line in userbreaks:
        broken_lines = wrap(line, var.max_chars)
        for new_line in broken_lines:
            lines.append(new_line)

    y = var.text_height
    for line in lines:
        canvas.drawString(var.dist_left, y, line)  # Escreve a linha por cima do modelo.
        y -= var.line_space  # Espaçamento entre as linhas

    canvas.save()  # Função I/O-intensiva para salvar o arquivo da memória para o armazenamento do computador.


def emit_from_source() -> None:
    """
    Função para emitir diversos certificados advindos de uma fonte de dados externa.
    """

    start_time = time()  # Horário do começo da emissão.
    grupo = get_grupo(var.grupo_dir)

    # Abre a planilha no modo leitura otimizada para obter o número de linhas e então fechá-la.
    row_count_time = time()
    workbook = load_workbook(filename=var.data_dir, read_only=True)
    sheet = workbook.worksheets[0]
    row_count = sheet.max_row - 1  # Desconta os cabeçários
    workbook.close()
    print(f'Tempo necessário para calcular que há {row_count} linhas na planilha: {(time() - row_count_time):.2f} segundos.')

    var.max_progress = row_count
    if row_count > 1000 and var.max_processes != 1:
        # Sabendo o número total de linhas, divide-se o trabalho de carregá-las na memória através de multiprocessamento.
        quotient = floor(var.max_progress / var.max_processes)  # Número de linhas que cada processo irá carregar na memória.
        remainder = var.max_progress % var.max_processes  # Número de linhas a mais que o último processo carregará na memória.
        processes = []

        for i in range(var.max_processes):
            if i == range(var.max_processes)[-1]:  # Se estiver na última iteração, dê a sobra de linhas para o último processo.
                processes.append(SubLoader(var.data_dir, (i * quotient) + 1, quotient + remainder, i, var.shared_list, var.lock))

            elif i == 0:  # Primeira linha. Em todos os casos, deve-se adicionar mais 1, pois senão há overlapping
                processes.append(SubLoader(var.data_dir, 1, quotient, i, var.shared_list, var.lock))

            else:  # Senão, dê apenas o quociente da divisão para esses processos que não são o último.
                processes.append(SubLoader(var.data_dir, (i * quotient) + 1, quotient, i, var.shared_list, var.lock))

            processes[i].start()

        for i in range(var.max_processes):
            processes[i].join()  # Espera todos os processos carregarem as suas seções de dados na memória para então continuar o programa.

        df = concat(var.shared_list[:], ignore_index=True)  # Junta todas as partes num inteiro.
        var.shared_list = var.manager.list(range(var.max_processes))  # Deleta os fragmentos de dados após serem utilizados.

    else:
        df = load(var.data_dir, header=None, skiprows=1)

    threads = []
    emit_time = time()  # Usado para calcular o tempo total e velocidade média do processo de emissão.

    # Usados para calcular velocidade marginal.
    marginal_time = time()
    marginal_cert = 0
    j = 0

    # Isso é necessário para evitar crashes quando o usuário decide continuar mesmo faltando campos.
    # E para caso as palavras-chave estejam vazias.
    arguments = []
    if var.headers['name'] != '':
        arguments.append(lambda i: df.iloc[i, var.headers['name']])
    else:
        arguments.append(lambda i: 'ERRO')

    if var.headers['cpf'] != '':
        arguments.append(lambda i: df.iloc[i, var.headers['cpf']])
    else:
        arguments.append(lambda i: 'ERRO')

    if var.headers['cnpj'] != '':
        arguments.append(lambda i: df.iloc[i, var.headers['cnpj']])
    else:
        arguments.append(lambda i: 'ERRO')

    if var.headers['matr'] != '':
        arguments.append(lambda i: df.iloc[i, var.headers['matr']])
    else:
        arguments.append(lambda i: 'ERRO')

    if var.headers['clie'] != '':
        arguments.append(lambda i: df.iloc[i, var.headers['clie']])
    else:
        arguments.append(lambda i: 'ERRO')

    if var.headers['apol'] != '':
        arguments.append(lambda i: df.iloc[i, var.headers['apol']])
    else:
        arguments.append(lambda i: 'ERRO')

    if var.headers['capi'] != '':
        arguments.append(lambda i: df.iloc[i, var.headers['capi']])
    else:
        arguments.append(lambda i: 'ERRO')

    if var.headers['cobe'] != '' and var.headers['grup'] != '':
        arguments.append(lambda i: df.iloc[i, var.headers['cobe']])
        arguments.append(lambda i: '')
    elif var.headers['cobe'] != '' and var.headers['grup'] == '':
        arguments.append(lambda i: df.iloc[i, var.headers['cobe']])
        arguments.append(lambda i: '')
    elif var.headers['cobe'] == '' and var.headers['grup'] != '':
        arguments.append(lambda i: '')
        arguments.append(lambda i: grupo.loc[neat.grupo(df.iloc[i, var.headers['grup']])])
    else:
        arguments.append(lambda i: 'ERRO')
        arguments.append(lambda i: 'ERRO')

    while j < var.max_progress:
        needed = var.target_threads - active_count()  # Três threads estão sempre ativas: (MainThread, update_progressbar_daemon, e emit_from_source) por isso a opção mínima é quatro.
        if j + needed > var.max_progress:
            needed = var.max_progress - j

        elif needed <= 0:
            needed = 1

        if active_count() < var.max_threads:
            threads.extend(Thread(target=emit_singular, args=[arg(i) for arg in arguments]) for i in range(j, j + needed))

            for i in range(1, needed + 1):
                threads[-i].start()
            j += needed

        else:
            sleep(0.001)

        if marginal_cert != var.progress:
            try:
                print(f'Velocidade marginal: {(var.progress - marginal_cert) / (time() - marginal_time):07.2f} certificados por segundo. | Threads ativas: {active_count():03}. | Progresso: {var.progress:04} de {var.max_progress:04}. | Iniciados: {j:04}. | Threads requeridas: {needed}.')
            except ZeroDivisionError:
                print(f'Velocidade marginal: 9999.99 certificados por segundo. | Threads ativas: {active_count():03}. | Progresso: {var.progress:04} de {var.max_progress:04}. | Iniciados: {j:04}. | Threads requeridas: {needed}.')
            marginal_time = time()
            marginal_cert = var.progress

    for thread in threads:
        thread.join()

    del df

    var.emission_time = time() - start_time
    var.certificates_per_second = var.max_progress / var.emission_time

    var.progress = 0
    print(f'Tempo levado para salvar {var.max_progress} arquivos PDF da memória para o disco: {(time() - emit_time) / 60:.2f} minutos')
    print(f'Velocidade média: {var.certificates_per_second:.2f}')


var.py
from PySide6.QtWidgets import QFileDialog, QMessageBox, QApplication
from pdfrw.buildxobj import pagexobj, PdfDict
from reportlab.pdfgen.canvas import Canvas
from multiprocessing import cpu_count
from locale import setlocale, LC_ALL
from pdfrw import PdfReader
from yaml import safe_load
from os.path import exists
from datetime import date
from os import getcwd
from sys import argv


def critical_select_sheet(name: str) -> str | None:
    """
    Quando o programa falha em encontrar alguma planilha essencial que ele precisa para funcionar, ele exigirá que uma seja selecionada, ou terminará.
    """

    while True:
        warning = QMessageBox()
        warning.setWindowTitle('AVISO')
        warning.setIcon(QMessageBox.Icon.Warning)
        warning.setStandardButtons(QMessageBox.StandardButton.Ok | QMessageBox.StandardButton.Close)
        warning.setText(f'Arquivo contendo {name} não foi encontrado\n'
                        f'Clique em ok para selecioná-lo\n'
                        f'Ou clique em close para terminar o programa.')

        if warning.exec() == QMessageBox.StandardButton.Ok:
            file = QFileDialog.getOpenFileName()[0]

            if file[-4:].lower() in ('xlsx', '.xls', '.csv', 'xlsm'):
                return file.replace(work_directory, '')

        else:
            exit()


def critical_select_pdf() -> str | None:
    """
    Quando o programa falhar em encontrar o arquivo PDF modelo, ele exigirá que um seja selecionado ou terminará.
    """

    while True:
        warning = QMessageBox()
        warning.setWindowTitle('AVISO')
        warning.setIcon(QMessageBox.Icon.Warning)
        warning.setStandardButtons(QMessageBox.StandardButton.Ok | QMessageBox.StandardButton.Close)
        warning.setText(f'O PDF modelo não foi encontrado\n'
                        f'Clique em ok para selecioná-lo\n'
                        f'Ou clique em close para terminar o programa.')

        if warning.exec() == QMessageBox.StandardButton.Ok:
            file = QFileDialog.getOpenFileName()[0]

            if file[-4:].lower() == '.pdf':
                return file.replace(work_directory, '')

        else:
            exit()


def get_pdf_template() -> PdfDict:  # Abre o modelo PDF. (Isso tem que ser feito aqui para poder mudá-lo posteriormente nas configurações)
    """
    Carrega o modelo PDF na memória e o retorna, onde ele poderá ser utilizado para construir outros PDFs a partir de si e de um texto.
    """

    template = PdfReader(template_dir, decompress=False)
    return pagexobj(template.pages[0])  # Retorna um objeto do PDF modelo.


built_in_fonts = Canvas('').getAvailableFonts()  # Todas as fontes já embutidas no ReportLab
setlocale(LC_ALL, 'pt-BR.UTF-8')  # Localização é usado para formatar dinheiro quando há campo capital

app = QApplication(argv)  # É necessário que o QApplication seja inicializado antes das possíveis janelas de erro e QDialogs que o podem vir a usá-lo.

progress                = 0
max_progress            = 1
certificates_per_second = 0
emission_time           = 0
work_directory          = getcwd().replace('\\', '/') + '/'  # Diretório onde está sendo executado o programa para que ele possa salvar outros diretórios de maneira mais legível.
load_problem            = False  # Quando há algum problema de carregamento, o programa salva imediatamente após concluir um carregamento correto.
abort_emission          = False  # Usado para dar a opção de abortar a emissão quando a planilha está faltando colunas críticas.
headers = ''

# Essas variáveis são inicializadas aqui, mas seus valores são atribuídos no módulo main.
manager     = ''
shared_list = ''
lock        = ''

if not exists('dados/init.yaml'):
    open('dados/init.yaml', 'x').close()

with open('dados/init.yaml', 'r') as file:
    dic = safe_load(file.read())

# Código boilerplate para carregamento do arquivo init.yaml e possíveis erros que podem vir a acontecer.
try:
    max_threads = dic['max_threads']
except (KeyError, TypeError):
    print('Erro ao carregar número máximo de threads, assumindo valor padrão de 150.')
    max_threads = 150
    load_problem = True

try:
    max_processes = dic['max_processes']
except (KeyError, TypeError):
    print(f'Erro ao carregar número máximo de processos, assumindo valor padrão de {cpu_count()}.')
    max_processes = cpu_count()
    load_problem = True

try:
    target_threads = dic['target_threads']
except (KeyError, TypeError):
    print('Erro ao carregar número alvo de threads, assumindo valor padrão de 60.')
    target_threads = 60
    load_problem = True

try:
    data_dir = dic['data_dir']
except (KeyError, TypeError):
    print('Erro ao carregar diretório do arquivo padrão de emissão, pois não foi guardado. Deselecionando-o.')
    data_dir = ''
    load_problem = True
finally:
    if data_dir != '' and not exists(data_dir):
        print('Erro ao carregar diretório do arquivo de inicialização, pois ele não existe. Deselecionando-o.')
        data_dir = ''
        load_problem = True

try:
    output_dir = dic['output_dir']
except (KeyError, TypeError):
    print('Erro ao carregar diretório de saída, pois não foi guardado. Deselecionando-o.')
    output_dir = ''
    load_problem = True
finally:
    if output_dir != '' and not exists(output_dir):
        print('Erro ao carregar diretório de saída, pois ele não existe. Deselecionando-o.')
        output_dir = ''
        load_problem = True

try:
    start_period = dic['start_period']
except (KeyError, TypeError):
    print('Erro ao carregar data inicial de vigência, assumindo data padrão de 01/01/2023. ')
    start_period = date(2001, 1, 1)
    load_problem = True

try:
    end_period = dic['end_period']
except (KeyError, TypeError):
    print('Erro ao carregar data final de vigência, assumindo data padrão de 01/01/2023. ')
    end_period = date(2001, 1, 1)
    load_problem = True

try:
    grupo_dir = dic['grupo_dir']
except (KeyError, TypeError):
    print('Erro ao carregar arquivo de códigos de grupos, pois não foi guardado. Começando processo para selecioná-los.')
    load_problem = True
    grupo_dir = critical_select_sheet('grupos')
finally:
    if not exists(grupo_dir):
        print('Erro ao carregar arquivo de códigos de grupos, pois ele não existe. Começando processo para selecioná-lo.')
        load_problem = True
        grupo_dir = critical_select_sheet('grupos')

try:
    template_dir = dic['template_dir']
except (KeyError, TypeError):
    print('Erro ao carregar arquivo modelo PDF, pois não foi guardado. Começando processo para selecioná-lo.')
    load_problem = True
    template_dir = critical_select_pdf()
finally:
    if not exists(template_dir):
        print('Erro ao carregar arquivo modelo PDF, pois ele não existe. Começando processo para selecioná-lo.')
        load_problem = True
        template_dir = critical_select_pdf()

try:
    text_font = dic['text_font']
except (KeyError, TypeError):
    print('Erro ao carregar fonte do texto, assumindo valor padrão de Times-Roman')
    text_font = 'Times-Roman'
    load_problem = True

try:
    text_size = dic['text_size']
except (KeyError, TypeError):
    print('Erro ao carregar tamanho do texto, assumindo valor padrão de 12')
    text_size = 12
    load_problem = True

try:
    dist_left = dic['dist_left']
except (KeyError, TypeError):
    print('Erro ao carregar espaçamento à esquerda, assumindo valor padrão de 80.')
    dist_left = 80
    load_problem = True

try:
    text_height = dic['text_height']
except (KeyError, TypeError):
    print('Erro ao carregar altura do texto, assumindo valor padrão de 620.')
    text_height = 620
    load_problem = True

try:
    line_space = dic['line_space']
except (KeyError, TypeError):
    print('Erro ao carregar espaçamento entre as linhas, assumindo valor padrão de 15.')
    line_space = 15
    load_problem = True

try:
    max_chars = dic['max_chars']
except (KeyError, TypeError):
    print('Erro ao carregar número de caracteres antes de quebrar a linha, assumindo valor padrão de 85.')
    max_chars = 85
    load_problem = True

try:
    base_text = dic['base_text']
except (KeyError, TypeError):
    print('Erro ao carregar texto base padrão, continuando sem ele.')
    base_text = ''
    load_problem = True

try:
    name_keywords = dic['name_keywords']
except (KeyError, TypeError):
    print('Erro ao carregar palavras-chave para nome, assumindo valor padrão de "NOME".')
    name_keywords = ['NOME']
    load_problem = True

try:
    cpf_keywords = dic['cpf_keywords']
except (KeyError, TypeError):
    print('Erro ao carregar palavras-chave para CPF, assumindo valor padrão de "CPF".')
    cpf_keywords = ['CPF']
    load_problem = True

try:
    cnpj_keywords = dic['cnpj_keywords']
except (KeyError, TypeError):
    print('Erro ao carregar palavras-chave para CNPJ, assumindo valor padrão de "CNPJ".')
    cnpj_keywords = ['CNPJ']
    load_problem = True

try:
    matricula_keywords = dic['matricula_keywords']
except (KeyError, TypeError):
    print('Erro ao carregar palavras-chave para matrícula, assumindo valor padrão de "MATRICULA".')
    matricula_keywords = ['MATRICULA']
    load_problem = True

try:
    cliente_keywords = dic['cliente_keywords']
except (KeyError, TypeError):
    print('Erro ao carregar palavras-chave para cliente, assumindo valor padrão de "CLIENTE".')
    cliente_keywords = ['CLIENTE']
    load_problem = True

try:
    apolice_keywords = dic['apolice_keywords']
except (KeyError, TypeError):
    print('Erro ao carregar palavras-chave para apólice, assumindo valor padrão de "APOLICE".')
    apolice_keywords = ['APOLICE']
    load_problem = True

try:
    capital_keywords = dic['capital_keywords']
except (KeyError, TypeError):
    print('Erro ao carregar palavras-chave para capital, assumindo valor padrão de "CAPITAL".')
    capital_keywords = ['CAPITAL']
    load_problem = True

try:
    cobertura_keywords = dic['cobertura_keywords']
except (KeyError, TypeError):
    print('Erro ao carregar palavras-chave para cobertura, assumindo valor padrão de "COBERTURA".')
    cobertura_keywords = ['COBERTURA']
    load_problem = True

try:
    grupo_keywords = dic['grupo_keywords']
except (KeyError, TypeError):
    print('Erro ao carregar palavras-chave para GRUPOS, assumindo valor padrão de "grupo".')
    grupo_keywords = ['grupo']
    load_problem = True

template_obj = get_pdf_template()
